File: 3116-kth-smallest-amount-with-single-denomination-combination/3116-kth-smallest-amount-with-single-denomination-combination.java
class Solution {

    long gcd(long a, long b) {
        while (b != 0) {
            long t = a % b;
            a = b;
            b = t;
        }
        return a;
    }

    long lcm(long a, long b) {
        return a / gcd(a, b) * b;
    }

    long count(long x, int[] coins) {
        int n = coins.length;
        long result = 0;

        for (int mask = 1; mask < (1 << n); mask++) {
            long mult = 1;
            int bits = 0;
            boolean valid = true;

            for (int i = 0; i < n; i++) {
                if ((mask & (1 << i)) != 0) {
                    bits++;

                    mult = lcm(mult, coins[i]);

                    if (mult > x) {
                        valid = false;
                        break;
                    }
                }
            }

            if (!valid)
                continue;

            long ways = x / mult;

            if (bits % 2 == 1)
                result += ways;
            else
                result -= ways;
        }

        return result;
    }

    public long findKthSmallest(int[] coins, int k) {
        long min = coins[0];

        for (int c : coins) {
            min = Math.min(min, c);
        }

        long low = 1;
        long high = min * (long) k;

        while (low < high) {
            long mid = low + (high - low) / 2;

            if (count(mid, coins) >= k)
                high = mid;
            else
                low = mid + 1;
        }

        return low;
    }
}

File: 3116-kth-smallest-amount-with-single-denomination-combination/3116-kth-smallest-amount-with-single-denomination-combination.js
/**
 * @param {number[]} coins
 * @param {number} k
 * @return {number}
 */
var findKthSmallest = function(coins, k) {

    const n = coins.length;

    function gcd(a, b) {
        while (b !== 0) {
            let t = a % b;
            a = b;
            b = t;
        }
        return a;
    }

    function lcm(a, b) {
        return a / gcd(a, b) * b;
    }

    function count(x) {
        let result = 0;

        for (let mask = 1; mask < (1 << n); mask++) {
            let mult = 1;
            let bits = 0;
            let valid = true;

            for (let i = 0; i < n; i++) {
                if (mask & (1 << i)) {
                    bits++;

                    mult = lcm(mult, coins[i]);

                    if (mult > x) {
                        valid = false;
                        break;
                    }
                }
            }

            if (!valid) continue;

            const ways = Math.floor(x / mult);

            if (bits % 2 === 1)
                result += ways;
            else
                result -= ways;
        }

        return result;
    }

    let low = 1;
    let high = Math.min(...coins) * k;

    while (low < high) {
        const mid = Math.floor((low + high) / 2);

        if (count(mid) >= k)
            high = mid;
        else
            low = mid + 1;
    }

    return low;
};